### ISSUE TITLE
Should have NAT Traversal

### ISSUE THREAD

**Issue by whyrusleeping** [ipfs/go-ipfs#57]
Peers behind nats need a way for other peers in the network to connect to them. It would be nice to also have webrtc, but until a go library for that exists, this will have to do.

**Comment by jbenet** [comment]
# Alpha: NAT or Relay

NAT traversal is non trivial, the roadmap looks like this:

- [x] relay service for ICE comm -- our p2p/protocols/relay
- [x] allow ipfs nodes to find suitable relays -- i.e. nodes directly connected to the node we wish to reach (implemented with `(*IpfsDht).FindPeersConnectedToPeer`)
- [ ] use something to perform ICE (current candidate: `natty`, which gives us only `UDPAddr` pairs) 
- [ ] use reliable transport over UDP.

The blocker is reliable transport. Our options so far have been:
- `utp` - needs to be implemented robustly. `h2so5/utp` is not yet working, could go-wrap `libutp`.
- `udt` - go-wrapped, needs some ARM assembly 
- `sctp` - needs go-wrapping, should not have assembly.
- `quic` - needs lib-izing, and go-wrapping. 

Of these, `quic` is my favorite option for the future. It is a truly phenmenal transport, with tons of things done very, very well. However, though the dragon is ready to hatch from its egg, it must still be unearthed from the enormous Code Caverns, of the Chromium Mountains, a quest not to be taken lightly.

Reliable transport a really big deal for us, and hard to get right:

1. Buggy transports will cause us endless pain and suffering. They already have: I spent many hours debugging dht and bitswap problems only to find that utp was the problem all along.
2. "Congestion Control" and "Flow Control" are both very hard. It took the TCP, SCTP, UTP, and QUIC teams many months to perfect designs and implementations.
3. Low {memory, cpu} footprints, etc are hard to achieve well, while also achieving low bandwidth and latency overheads.

@whyrusleeping pre-empting your call for "Lets Implement UTP!", we should not roll our own right now-- it will take us _weeks_ to get reasonable robustness, speed, flow control, memory consumption, etc. We are good hackers, but it's taken the broader community _years_ to make the few implementations there are. this is not to be underestimated. I totally want to have native go implementations of these transports (see my related emails to go-nuts mailing list), but we cannot afford the implementation cost right now.

### another option TCP/UDP tunnel

And, frankly, I'm seriously considering doing the following horrible hack that might actually work __robustly__ and be much quicker to pull off that 

- use the `UDPAddr` pairs from natty, BUT!
- setup a TCP/UDP tunnel with our own NAT:
		```
		A-TCPConn <---> IPConn <---> UDPConn <--IPConn--> TCPConn-B
		``` 

This is a **mad science**(tm) hack that gets around the awful realities of the network and the awful realities of software: simple things outside of specific rails is too fucking hard to do.

Looking at http://golang.org/pkg/net/#IPConn -- it should be possible, my concern is making sure `DialIP` doesn't interfere with other traffic:

Resources:
- https://godoc.org/golang.org/x/net/ipv4
- https://github.com/kdar/gorawtcpsyn/blob/master/main.go
- http://stackoverflow.com/questions/18427655/use-raw-sockets-in-go

### Relay past NAT -- easy cop-out for now

The easiest thing to do for now is to just relay _all_ traffic. Would get us around NAT, but would certainly increase load of nodes. 

We'll _have to_ relay some traffic, as not all NAT is hole punching. Sometimes we do have to use relaying. This could actually be done with Bitswap in mind, and be governed by the standard bitswap decision engine: relay for those nodes who are useful to us.

In lieu of using natty and hole-punching, we could start with the big guns (which will work for everything and we have to do anyway) and optimize later. My [Knuth](http://www-cs-faculty.stanford.edu/~uno/don.gif) VM is telling me to [optimize later](http://en.wikipedia.org/wiki/Program_optimization#When_to_optimize).

**Comment by whyrusleeping** [comment]
I think a combination of relays, and NAT hole punching (NAT-PMP, uPnP) could be our best option (at least short term). I found a PMP library: https://github.com/jackpal/go-nat-pmp, and there was a few promising uPnP libraries around, which should get us through a decent number of networks, and the rest can be relayed.

**Comment by briantigerchow** [comment]
> we could start with the big guns

the big guns being relay?

**Comment by jbenet** [comment]
> the big guns being relay?

yeah

**Comment by jbenet** [comment]
So this happened: https://groups.google.com/a/chromium.org/d/msg/proto-quic/MIogHT8wacA/69H8gchp0nMJ (QUIC)

**Comment by tv42** [comment]
Since I don't think I've seen this mentioned yet, and it's good overview:

https://tools.ietf.org/html/rfc5128
State of Peer-to-Peer (P2P) Communication across Network Address Translators (NATs)

**Comment by drwasho** [comment]
Any update on where you're at on this issue?

**Comment by whyrusleeping** [comment]
closing this for now, basic NAT traversal has been implemented as well as utp in dev0.4.0. Further discussion on relays and more advanced traversal should start a new issue

**Comment by drwasho** [comment]
Awesome, well done guys!

**Comment by myleshorton** [comment]
Hey @whyrusleeping and @jbenet can you describe the NAT traversal implementation in 0.4.0 or link to more details? Exciting stuff.

**Comment by Cole128** [comment]
@myleshorton @whyrusleeping @jbenet  Yes, I am interested in how it works for the purposes of my own decentralized application. My application would require IPFS anyway, and I am wondering if I can just piggyback on the NAT traversal that IPFS uses, and then just specify a different a different port?

**Comment by Cole128** [comment]
I just came back to this link again. Does anyone have an answer?

**Comment by Cole128** [comment]
@lidel libp2p led me [here](https://github.com/libp2p/js-libp2p/issues/104) which seems to say it is only kind of implemented now.

**Comment by Stebalien** [comment]
@Cole128 there are two implementations of libp2p, one in js and one in go (and a work in progress one in rust). The issue you linked was for js. NAT traversal works pretty well in go (although we're still missing a few pieces).

If you just want to piggyback on ipfs's NAT traversal, you may be interested in https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#ipfs-p2p

There's some improved documentation in #4894 but it's blocked on actually making this feature behave according to that documentation... (i.e., how one would expect).

If you're still having trouble with NAT traversal, you may also be interested in the circuit relay feature: https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#circuit-relay

**Comment by Cole128** [comment]
@Stebalien thanks a lot, I'll check all those links out.
